Add Collection API

Wauplin opened the pull request:
First draft to support the Collection API (implements https://github.com/huggingface/huggingface_hub/issues/1682).

### Create/get/update/delete collection:
- `get_collection`
- `create_collection`: title, description, namespace, private
- `update_collection_metadata`: title, description, position, private, theme
- `delete_collection`

It does not include a way to list/search collections. We can do that in a later PR once it's fully supported server-side (currently only possible to list your own collections).

###  Add/update/remove item from collection:
- `add_collection_item`: item id, item type, note
- `update_collection_item`: note, position
- `delete_collection_item`

### Usage

```py
>>> from huggingface_hub import get_collection
>>> collection = get_collection("TheBloke/recent-models-64f9a55bb3115b4f513ec026")
>>> collection.title
'Recent models'
>>> len(collection.items)
37
>>> collection.items[0]
CollectionItem: {
    {'_id': '6507f6d5423b46492ee1413e',
    'id': 'TheBloke/TigerBot-70B-Chat-GPTQ',
    'author': 'TheBloke',
    'item_type': 'model',
    'lastModified': '2023-09-19T12:55:21.000Z',
    (...)
}}
```

```py
>>> from huggingface_hub import create_collection

# Create collection
>>> collection = create_collection(
...     title="ICCV 2023",
...     description="Portfolio of models, papers and demos I presented at ICCV 2023",
... )

# Add collection with a note
>>> add_collection_item(
...     collection_slug="davanstrien/climate-64f99dc2a5067f6b65531bab",
...     item_id="datasets/climate_fever",
...     item_type="dataset"
...     note="This dataset adopts the FEVER methodology that consists of 1,535 real-world claims regarding climate-change collected on the internet."
... )
```
### TODO

- [x] merge [moon-landing/7536](https://github.com/huggingface/moon-landing/pull/7536) (internal)
- [ ] merge [moon-landing/7538](https://github.com/huggingface/moon-landing/pull/7538) (internal)
- [ ] add unit test
- [ ] add Collection API to package reference (methods + classes)
- [ ] add guide about Collection API
- [ ] document public API in hf-docs (different PR)

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/huggingface_hub/pr_1687). All of your documentation changes will be reflected on that endpoint.

osanseviero commented on docs/source/en/guides/collections.md:
```diff
@@ -0,0 +1,187 @@
+
+>>> collection = create_collection(title="OS Week Highlights - Sept 18 - 24", namespace="osanseviero")
+>>> collection.slug
+"osanseviero/os-week-highlights-sept-18-24-650bfed7f795a59f491afb80"
```
nice find :D

Wauplin replied:
Always try to highlight some cool stuff in the docs examples :hugs:

julien-c replied:
nice growth hack for Omar's profile:)

LysandreJik commented:
cc @vwxyzjn who was wondering whether this was possible or not

LysandreJik reviewed (approved):
Clean implementation, a welcome addition!

Maybe we could also eventually add support for that in `transformers`/`diffusers`/etc or at least show in the docs how to add to a collection a fresh model uploaded using `push_to_hub`

stevhliu commented on docs/source/en/guides/collections.md:
```diff
@@ -0,0 +1,187 @@
+rendered properly in your Markdown viewer.
+-->
+
+# Manage your collections
```
I think it may be more consistent to align the doc title with the title in the `toctree`; this makes it easier for users to use the same name to reference the doc

```suggestion
# Collections
```

stevhliu commented on docs/source/en/guides/collections.md:
```diff
@@ -0,0 +1,187 @@
+
+# Manage your collections
+
+A collection is a group of related items on the Hub (models, datasets, Spaces, papers) that are organized together on a same page. Collections can be useful in many use cases such as creating your own portfolio, bookmarking content in categories or presenting a curated list of items your want to share. Check out this [guide](https://huggingface.co/docs/hub/collections) to understand in more detai ...
```
```suggestion
A collection is a group of related items on the Hub (models, datasets, Spaces, papers) that are organized together on the same page. Collections are useful for creating your own portfolio, bookmarking content in categories, or presenting a curated list of items you want to share. Check out this [guide](https://huggingface.co/docs/hub/collections) to understand in more detail what collections are and how they look on the Hub.
```

stevhliu commented on docs/source/en/guides/collections.md:
```diff
@@ -0,0 +1,187 @@
+
+A collection is a group of related items on the Hub (models, datasets, Spaces, papers) that are organized together on a same page. Collections can be useful in many use cases such as creating your own portfolio, bookmarking content in categories or presenting a curated list of items your want to share. Check out this [guide](https://huggingface.co/docs/hub/collections) to understand in more detai ...
+
+Managing collections can be done in the browser directly. In this guide, we will focus on how to do it programmatically using `huggingface_hub`.
```
```suggestion
You can directly manage collections in the browser, but in this guide, we will focus on how to manage it programmatically.
```

stevhliu commented on docs/source/en/guides/collections.md:
```diff
@@ -0,0 +1,187 @@
+
+## Fetch a collection
+
+To fetch a collection, use [`get_collection`]. You can use it either on your own collections or any public one. To retrieve a collection, you must have its collection's `slug`. A slug is an identifier for a collection based on the title and a unique ID. You can find it in the URL of the collection page.
```
```suggestion
Use [`get_collection`] to fetch your collections or any public ones. You must have the collection's *slug* to retrieve a collection. A slug is an identifier for a collection based on the title and a unique ID. You can find the slug in the URL of the collection page.
```

stevhliu commented on docs/source/en/guides/collections.md:
```diff
@@ -0,0 +1,187 @@
+To fetch a collection, use [`get_collection`]. You can use it either on your own collections or any public one. To retrieve a collection, you must have its collection's `slug`. A slug is an identifier for a collection based on the title and a unique ID. You can find it in the URL of the collection page.
+
+<div class="flex justify-center">
+    <img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/hfh_collection_slug.png"/>
```
Minor detail, but I think it'd be nice to highlight/underline the slug in the URL to immediately draw the user's eyes to it.

Wauplin replied:
Good idea. I updated the remote image: https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/hfh_collection_slug.png

stevhliu commented on docs/source/en/guides/collections.md:
```diff
@@ -0,0 +1,187 @@
+    <img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/hfh_collection_slug.png"/>
+</div>
+
+Here the slug is `"TheBloke/recent-models-64f9a55bb3115b4f513ec026"`. Let's fetch the collection:
```
```suggestion
Let's fetch the collection with, `"TheBloke/recent-models-64f9a55bb3115b4f513ec026"`:
```

stevhliu commented on docs/source/en/guides/collections.md:
```diff
@@ -0,0 +1,187 @@
+
+The [`Collection`] object returned by [`get_collection`] contains:
+- high-level metadata: `slug`, `owner`, `title`, `description`, etc.
+- a list of [`CollectionItem`] objects. Each item represents a model, a dataset, a Space or a paper.
```
```suggestion
- a list of [`CollectionItem`] objects; each item represents a model, a dataset, a Space, or a paper.
```

stevhliu commented on docs/source/en/guides/collections.md:
```diff
@@ -0,0 +1,187 @@
+- high-level metadata: `slug`, `owner`, `title`, `description`, etc.
+- a list of [`CollectionItem`] objects. Each item represents a model, a dataset, a Space or a paper.
+
+All items of a collection are guaranteed to have:
```
```suggestion
All collection items are guaranteed to have:
```

stevhliu commented on docs/source/en/guides/collections.md:
```diff
@@ -0,0 +1,187 @@
+
+All items of a collection are guaranteed to have:
+- a unique `item_object_id`: this is the id of the collection item in the database
+- an `item_id`: this is the id on the Hub of the underlying item (model, dataset, Space, paper). It is not necessarily unique! only the `item_id`/`item_type` pair is unique.
```
```suggestion
- an `item_id`: this is the id on the Hub of the underlying item (model, dataset, Space, paper); it is not necessarily unique, and only the `item_id`/`item_type` pair is unique
```

stevhliu commented on docs/source/en/guides/collections.md:
```diff
@@ -0,0 +1,187 @@
+All items of a collection are guaranteed to have:
+- a unique `item_object_id`: this is the id of the collection item in the database
+- an `item_id`: this is the id on the Hub of the underlying item (model, dataset, Space, paper). It is not necessarily unique! only the `item_id`/`item_type` pair is unique.
+- an `item_type`: model, dataset, Space, paper.
```
```suggestion
- an `item_type`: model, dataset, Space, paper
```

stevhliu commented on docs/source/en/guides/collections.md:
```diff
@@ -0,0 +1,187 @@
+- a unique `item_object_id`: this is the id of the collection item in the database
+- an `item_id`: this is the id on the Hub of the underlying item (model, dataset, Space, paper). It is not necessarily unique! only the `item_id`/`item_type` pair is unique.
+- an `item_type`: model, dataset, Space, paper.
+- the `position` of the item in the collection. Position can be updated to re-organize your collection (see [`update_collection_item`] below)
```
```suggestion
- the `position` of the item in the collection, which can be updated to reorganize your collection (see [`update_collection_item`] below)
```

stevhliu commented on docs/source/en/guides/collections.md:
```diff
@@ -0,0 +1,187 @@
+- an `item_type`: model, dataset, Space, paper.
+- the `position` of the item in the collection. Position can be updated to re-organize your collection (see [`update_collection_item`] below)
+
+A `note` can also be attached to the item. This is useful to add additional information about the item (e.g. a comment, a link to a blog post, etc.). If an item doesn't have a note, the attribute still exists with a `None` value.
```
Maybe link to how you can attach a `note`?

```suggestion
A `note` can also be attached to the item. This is useful to add additional information about the item (a comment, a link to a blog post, etc.). The attribute still has a `None` value if an item doesn't have a note.
```

stevhliu commented on docs/source/en/guides/collections.md:
```diff
@@ -0,0 +1,187 @@
+
+## Create a new collection
+
+Now that we know how to get a [`Collection`], let's create our own! Use [`create_collection`] with a title and optionally a description.
```
```suggestion
Now that we know how to get a [`Collection`], let's create our own! Use [`create_collection`] with a title and description.
```

stevhliu commented on docs/source/en/guides/collections.md:
````diff
@@ -0,0 +1,187 @@
+"username"
+```
+
+To create a collection on an organization page, pass `namespace="my-cool-org"` when creating the collection. Finally, you can also create private collections by passing `private=True`.
````
I think this paragraph makes more sense if you put it at the beginning of the section when you're describing `create_collection`

stevhliu commented on docs/source/en/guides/collections.md:
```diff
@@ -0,0 +1,187 @@
+
+## Manage items in a collection
+
+Now that we have a [`Collection`], we want to add items to it and organize them.
```
Nice job adding a sentence here describing the section instead of leaving it empty!

stevhliu commented on docs/source/en/guides/collections.md:
````diff
@@ -0,0 +1,187 @@
+>>> add_collection_item(collection.slug, item_id="warp-ai/wuerstchen", item_type="space") # same item_id, different item_type
+```
+
+If an item already exists in a collection (i.e. same `item_id`/`item_type` pair), an HTTP 409 error will be raised. You can choose to ignore this error by setting `exists_ok=True`.
````
```suggestion
If an item already exists in a collection (same `item_id`/`item_type` pair), an HTTP 409 error will be raised. You can choose to ignore this error by setting `exists_ok=True`.
```

stevhliu commented on docs/source/en/guides/collections.md:
````diff
@@ -0,0 +1,187 @@
+... )
+```
+
+### Re-order items
````
```suggestion
### Reorder items
```

stevhliu commented on docs/source/en/guides/collections.md:
```diff
@@ -0,0 +1,187 @@
+
+### Re-order items
+
+Items in a collection are ordered. The order is determined by the `position` attribute of each item. By default, items are ordered by appending new items at the end of the collection. You can update the ordering using [`update_collection_item`] the same way you would add a note.
```
```suggestion
Items in a collection are ordered. The order is determined by the `position` attribute of each item. By default, items are ordered by appending new items at the end of the collection. You can update the order using [`update_collection_item`] the same way you would add a note.
```

stevhliu commented on docs/source/en/guides/collections.md:
```diff
@@ -0,0 +1,187 @@
+
+### Remove items
+
+Finally you can also remove an item using [`delete_collection_item`].
```
```suggestion
Finally, you can also remove an item using [`delete_collection_item`].
```

stevhliu commented on docs/source/en/package_reference/collections.md:
```diff
@@ -0,0 +1,25 @@
+- Update an item in a collection: [`update_collection_item`]
+- Remove an item from a collection: [`delete_collection_item`]
+
+## Data structures
```
```suggestion
```

stevhliu reviewed (approved):
Super nice guide! 💯

Wauplin commented:
Thank you so much for the guide review @stevhliu ! :heart: I've addressed all of your comments which is always a great improvement to the guide :tada:

Wauplin commented:
Server-side PRs are not all merged and tests are green :heavy_check_mark:.

Still having a code quality error but it's been solved in https://github.com/huggingface/huggingface_hub/pull/1693 (`types-requests` got updated) and in PR documentation which is being solved in https://github.com/huggingface/doc-builder/pull/405. Both are not related to this PR.

So thanks @LysandreJik and @stevhliu for the reviews, I think we're good to merge now! :tada: :rocket:

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/huggingface_hub/pr_1687). All of your documentation changes will be reflected on that endpoint.

davanstrien commented:
Would adding a `collection_url`/`url` attribute to the `Collection` object make sense? I realise this can easily be got from the slug i.e. `f"https:huggingface.co/collections/{collection.slug}"` but maybe it's nice to make it easier for a user to quickly access that so they can check their collection via the Hub more quickly?

Wauplin commented:
> Would adding a collection_url/url attribute to the Collection object make sense? I realise this can easily be got from the slug i.e. f"https:huggingface.co/collections/{collection.slug}" but maybe it's nice to make it easier for a user to quickly access that so they can check their collection via the Hub more quickly?

Yes that makes a lot of sense for usability, thanks for the suggestion @davanstrien. I opened https://github.com/huggingface/huggingface_hub/pull/1695 to add it.

julien-c commented:
maybe pin this as instruction in the GH issues and/or even in the README! i think many community members will be interested
